### PR TITLE
Large last written lsn cache

### PR DESCRIPTION
**Pull request by knizhnik**
Maintain cache of last written LSN for each relation segment (8 Mb).
Unlike #167, it allows to handle updates of multiple relations (i.e. table with large number of indexes)
and parallel execution of inserts and selects for large append-only table.

### CODE REVIEW

File: contrib/neon/pagestore_smgr.c
Comment (hlinnaka) [inline contrib/neon/pagestore_smgr.c:799]: typo:
```suggestion
	 * relation. Currently, we don't call SetLastWrittenLSN() when a new
```

File: src/backend/access/transam/xlog.c
Comment (hlinnaka) [inline src/backend/access/transam/xlog.c:8916]: rnode isn't enough to uniquely identify a relation. If you have two tables with same relfilenode, what happens?

Also, what about FSM and VM forks?

File: contrib/neon/pagestore_smgr.c
Comment (hlinnaka) [inline contrib/neon/pagestore_smgr.c:1441]: Hmm, so you don't remove the cache entries for the truncated blocks? Is that ok?

File: src/backend/access/transam/xlog.c
Comment (hlinnaka) [inline src/backend/access/transam/xlog.c:8902]: Scanning the hash table seems expensive. How about tracking the max of all cached LSNs a separate variable, like `maxLastWrittenLsn`?

File: contrib/neon/pagestore_smgr.c
Comment (hlinnaka) [inline contrib/neon/pagestore_smgr.c:89]: It's a bit ugly to use this value because 0xFFFFFFFE is a valid block number. You will probably hit other problems before you hit that limit, but still. Could use 0xFFFFFFFF, i.e. InvalidBlockNumber instead.

File: src/backend/access/transam/xlog.c
Comment (knizhnik) [reply src/backend/access/transam/xlog.c:8916]: I have explained it in comment to LastWrittenLsnCacheKey struct: I just want to make key shorter.
Probabiity of conflict (i.e. that two different relations from different databases has the same oid and are together updated) is relatively small. Concern FSM/VM forks, it is even better that them are placed in the same backet, because them are accessed together.

In  any case: this mapping is not required to be unique: we store max LSN. This is why it is possible to store single LSN for range of pages (and forks/relations).

File: contrib/neon/pagestore_smgr.c
Comment (knizhnik) [reply contrib/neon/pagestore_smgr.c:1441]: I think that it is ok: LRU replacement algorithm is used to avoid cache overflow.
And if OID is reused for some other relation, then - see above: nothing critical can happen because of max(relid) assignment.

File: src/backend/access/transam/xlog.c
Comment (knizhnik) [reply src/backend/access/transam/xlog.c:8902]: Sorry, but `GetLastWrittenLsn`  is called with relnode=InvalidOid only by `zenith_dbsize`. It seems to be not very frequently used  and performance critic al request so I decided to not pay extra coast for mainaining maximum.
Especially it is not so clear how to adjust it when victim is evicted from the cache.

File: src/backend/access/transam/xlog.c
Comment (aome510) [inline src/backend/access/transam/xlog.c:783]: It seems that these variables are protected by `LastWrittenLsnLock` as the lock is acquired at the start of `GetLastWrittenLSN` and `SetLastWrittenLSN`. Can we add a comment to describe this protection?

File: src/backend/access/transam/xlog.c
Comment (aome510) [inline src/backend/access/transam/xlog.c:808]: nit: add lock protection note

File: src/backend/access/transam/xlog.c
Comment (aome510) [inline src/backend/access/transam/xlog.c:776]:
```suggestion
	 * Maximal last written LSN for pages not present in lastWrittenLsnCache
```

File: src/backend/access/transam/xlog.c
Comment (knizhnik) [reply src/backend/access/transam/xlog.c:8902]: I do not see contradiction here. `GetLastWrittenLsn(InvaidOid)` calculates maximum among cached LSNs and `maxLastWrittenLsn` (maximal LSN for all non cached relations/buckets). So it actually returns what was originally stored in `lastWrittenLsn`.
`GetLastWrittenLsn(InvaidOid)`  is used now only by zenith_dbsize.

`SetLastWrittenLsn` i called by  createdb and dbase_redo. It updates only `maxLastWrittenLsn` . It can be that `maxLastWrittenLsn`  is less (or greater) than LSNs stored in the cache. It is ok, as far as we i any case take maximum among all of them.

File: src/backend/access/transam/xlog.c
Comment (aome510) [inline src/backend/access/transam/xlog.c:816]:
```suggestion
 * Size of the cache is limited by GUC variable lastWrittenLsnCacheSize ("lsn_cache_size"),
 * pages are replaced using LRU algorithm, based on L2-list.
```

**Review by aome510 — APPROVED** [review]
LGTM beside minor typo.

I did some comparison with the new [WAL backpressure tests](https://github.com/neondatabase/neon/pull/1919). The previous comparison I did for #167: https://github.com/neondatabase/postgres/pull/167#issuecomment-1160659340.

## Result

- `main` (https://app.circleci.com/pipelines/github/neondatabase/neon/7341/workflows/e7acd0ce-c7d5-491a-b2cd-babef720a7d6/jobs/75267)

<details>

```
test_heavy_write_workload[vanilla-10-5-5].run_duration: 41.399 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_max: 0.009 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_avg: 0.001 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_stdev: 0.002 s
test_heavy_write_workload[neon_off-10-5-5].run_duration: 226.222 s
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_max: 4 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_avg: 1 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_stdev: 1 MB
test_heavy_write_workload[neon_off-10-5-5].read_latency_max: 0.201 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_avg: 0.040 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_stdev: 0.031 s
test_heavy_write_workload[neon_on-10-5-5].run_duration: 244.264 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_max: 0.315 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_avg: 0.047 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_stdev: 0.051 s
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_max: 4 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_avg: 1 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_stdev: 1 MB
test_pgbench_simple_update_workload[vanilla-45-100].run_duration: 45.835 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_max: 0.058 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_avg: 0.002 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_stdev: 0.009 s
test_pgbench_simple_update_workload[neon_off-45-100].run_duration: 47.943 s
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_max: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_avg: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_max: 0.046 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_avg: 0.004 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_stdev: 0.007 s
test_pgbench_simple_update_workload[neon_on-45-100].run_duration: 49.697 s
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_max: 1 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_avg: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_max: 0.117 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_avg: 0.009 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_stdev: 0.018 s
test_pgbench_intensive_init_workload[vanilla-1000].run_duration: 114.447 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_max: 0.014 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_avg: 0.008 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_stdev: 0.002 s
test_pgbench_intensive_init_workload[neon_off-1000].run_duration: 61.270 s
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_max: 15 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_avg: 2 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_stdev: 3 MB
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_max: 3.062 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_avg: 0.880 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_stdev: 0.536 s
test_pgbench_intensive_init_workload[neon_on-1000].run_duration: 103.561 s
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_max: 15 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_avg: 10 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_stdev: 5 MB
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_max: 17.934 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_avg: 9.452 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_stdev: 5.661 s
```

</details>

- `main` + this patch (https://app.circleci.com/pipelines/github/neondatabase/neon/7327/workflows/4dfdc450-1f44-4c6a-85a6-bdae1a43b3b9/jobs/75101)

<details>

```
test_heavy_write_workload[vanilla-10-5-5].run_duration: 39.048 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_max: 0.007 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_avg: 0.001 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_stdev: 0.001 s
test_heavy_write_workload[neon_off-10-5-5].run_duration: 211.563 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_max: 0.944 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_avg: 0.075 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_stdev: 0.158 s
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_max: 17 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_avg: 14 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_stdev: 3 MB
test_heavy_write_workload[neon_on-10-5-5].run_duration: 222.842 s
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_max: 17 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_avg: 15 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_stdev: 2 MB
test_heavy_write_workload[neon_on-10-5-5].read_latency_max: 0.678 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_avg: 0.069 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_stdev: 0.132 s
test_pgbench_simple_update_workload[vanilla-45-100].run_duration: 45.166 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_max: 0.061 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_avg: 0.003 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_stdev: 0.013 s
test_pgbench_simple_update_workload[neon_off-45-100].run_duration: 46.441 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_max: 0.049 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_avg: 0.008 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_stdev: 0.011 s
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_max: 1 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_avg: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].run_duration: 47.743 s
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_max: 1 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_avg: 1 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_max: 0.119 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_avg: 0.012 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_stdev: 0.023 s
test_pgbench_intensive_init_workload[vanilla-1000].run_duration: 119.392 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_max: 0.233 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_avg: 0.010 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_stdev: 0.021 s
test_pgbench_intensive_init_workload[neon_off-1000].run_duration: 60.566 s
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_max: 12 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_avg: 2 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_stdev: 2 MB
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_max: 0.647 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_avg: 0.181 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_stdev: 0.140 s
test_pgbench_intensive_init_workload[neon_on-1000].run_duration: 108.036 s
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_max: 15 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_avg: 11 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_stdev: 5 MB
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_max: 2.113 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_avg: 0.251 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_stdev: 0.378 s
```

</details>

So far, at least for `intensive_init_workload`, this patch is not as good as #167. That's ok because #167 is designed for that workload anyhow. Both patches are still better than what we currently have in the `main` branch.

**Comment by knizhnik** [comment]
> > There is correspondent PR in neon repository with test, illustrating effect of this cache, but this test depends on test_backpressure branch.
> 
> You mean [neondatabase/neon#1949](https://github.com/neondatabase/neon/pull/1949)?

Yes

> 
> > So nay be we should first commit backpressure tests
> 
> FYI, the backpressure tests are merged into main ([neondatabase/neon#1919](https://github.com/neondatabase/neon/pull/1919)).

Thank you. I rebased #1949 with main.

**Comment by aome510** [comment]
@knizhnik can we merge it into `master`? Or are u still waiting for another approval?

**Comment by knizhnik** [comment]
> @knizhnik can we merge this into `master`? Or are u still waiting for another approval?

If @hlinnaka  has not objection then I will merge it in main.

File: src/backend/access/transam/xlog.c
Comment (hlinnaka) [reply src/backend/access/transam/xlog.c:190]: Also, if you're worried about performance, you could use lib/simplehash.h instead of dynahash. It's a bit faster. I don't think it matters in practice though.

**Review by hlinnaka — APPROVED** [review]
Left some more comments, but feel free to merge when you think it's ready.

File: src/backend/access/transam/xlog.c
Comment (knizhnik) [reply src/backend/access/transam/xlog.c:190]: > Seems a bit weird. If you have two databases it's very common to have relations with the same relfilenode in both. All the catalogs would start with the same relfilenode, for example.

> You're right that it's not a correctness problem, since the LSN stored is a conservative "max" value. But "false sharing" like that is weird. At the very least, this needs a comment to explain why it's OK.It would be very straightforward to avoid by using the whole RelFileNode and ForkNumber instead, and that would require less explaining too.

> You could store just the hash of RelFileNode and ForkNumber if you're worried about the cache size.

Well, may be my attempt to make hash calculation and key comparison a little bit faster by using only relfilenode is bad idea because of false sharing. My assumption were the following:
1. I think that very few free tier users will create a lot of databases. I actually not sure if creation of database is ever possible with current users API.
2. Yes, catalog table should have the same relfilenode among all databases. But I think that it is more important to separate rarely changed catalog tables fro user tables, rather than catalog tables of different databases. Yes, in principle it is possible that in one database catalog is actively updated (for example because of temp tables creation), and in others - not. In this case false sharing may be a problem.
3. I am intentionally ignore fork, because pages of main fork and FSM/VM forks are used to be updated together with correspondent pages of main fork. 

But if you insist, I can add full tag to cache entry.

**Pull request merged**
